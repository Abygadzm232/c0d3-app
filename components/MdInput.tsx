import React, { useState, useEffect, useRef } from 'react'
import Markdown from 'markdown-to-jsx'
import { colors } from './theme/colors'
import { Nav } from 'react-bootstrap'
import noop from '../helpers/noop'
import styles from '../scss/mdInput.module.scss'

type MdInputProps = {
  onChange?: Function
  placeHolder?: string
  value?: string
  bgColor?: 'white' | 'none'
}

const autoSize = (el: HTMLTextAreaElement) => {
  // Set height auto to shrink size on delete
  el.style.height = 'auto'
  // Set height to match scrollHeight (+2 prevents scroll bar from showing)
  el.style.height = el.scrollHeight + 2 + 'px'
}

export const MdInput: React.FC<MdInputProps> = ({
  bgColor = 'none',
  onChange = noop,
  placeHolder = 'Type something...',
  value = ''
}) => {
  const [preview, setPreview] = useState<boolean>(false)
  const textareaRef = useRef<HTMLTextAreaElement>(null)
  const mouseDownHeightRef = useRef<number | null>(null)
  const isMountedRef = useRef(false)
  const [height, setHeight] = useState<number | null>(null)

  useEffect(() => {
    // Focus when returning from preview mode after component has mounted
    if (!preview && isMountedRef.current) textareaRef.current?.focus()

    // Ref is used to prevent an unnecessary immediate re-render
    isMountedRef.current = true
  }, [preview])

  useEffect(() => {
    // Global event listener to capture mouse releases that may happen outside of the component
    const updateHeight = () => {
      if (!mouseDownHeightRef.current) return

      // Only update height if it changed
      if (
        textareaRef.current &&
        mouseDownHeightRef.current !== textareaRef.current.clientHeight
      )
        setHeight(textareaRef.current.clientHeight)

      // Always reset mouseDown ref
      mouseDownHeightRef.current = null
    }
    window.addEventListener('mouseup', updateHeight, false)

    return () => window.removeEventListener('mouseup', updateHeight, false)
  }, [])

  const previewBtnColor = preview ? 'black' : 'lightgrey'
  const writeBtnColor = preview ? 'lightgrey' : 'black'

  return (
    <div style={{ backgroundColor: bgColor }}>
      <Nav
        variant="tabs"
        defaultActiveKey="Write"
        activeKey={preview ? 'Preview' : 'Write'}
        onSelect={selectedKey => setPreview(selectedKey === 'Preview')}
      >
        <Nav.Item>
          <Nav.Link style={{ color: colors[writeBtnColor] }} eventKey="Write">
            Write
          </Nav.Link>
        </Nav.Item>
        <Nav.Item>
          <Nav.Link
            style={{ color: colors[previewBtnColor] }}
            eventKey="Preview"
          >
            Preview
          </Nav.Link>
        </Nav.Item>
      </Nav>
      {preview && (
        <>
          {value ? (
            <Markdown
              data-testid="markdown"
              style={{
                minHeight:
                  Math.min(textareaRef.current?.clientHeight || Infinity, 500) +
                  'px'
              }}
              className={`${styles['preview']}`}
            >
              {value}
            </Markdown>
          ) : (
            // Using textArea class so minHeight stays in sync
            <div className={styles['textarea']}>Nothing to preview</div>
          )}
          <hr />
        </>
      )}
      <textarea
        onMouseDown={() => {
          mouseDownHeightRef.current = textareaRef.current!.clientHeight
        }}
        ref={textareaRef}
        value={value}
        className={`${styles['textarea']}${preview ? ' d-none' : ''}`}
        onChange={e => {
          if (!height && textareaRef.current) autoSize(textareaRef.current)
          onChange(e.target.value)
        }}
        placeholder={placeHolder}
        style={height ? { height: height + 'px' } : undefined}
        data-testid="textbox"
      />
    </div>
  )
}
