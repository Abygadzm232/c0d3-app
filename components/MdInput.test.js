import React from 'react'
import { render, fireEvent, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { MdInput } from './MdInput'

const TestComponent = () => {
  const [testState, setTestState] = React.useState('')
  return <MdInput onChange={setTestState} value={testState} />
}

describe('MdInput Component', () => {
  test('Should render text onto textbox when user types', () => {
    const { container, queryByTestId } = render(<TestComponent />)
    const textbox = queryByTestId('textbox')
    fireEvent.change(textbox, {
      target: { value: 'Javascript' }
    })

    expect(textbox.value).toContain('Javascript')
    expect(container).toMatchSnapshot()
  })

  test('Should switch to Write mode when user clicks Write button', () => {
    const { container, getByRole, queryByTestId } = render(<MdInput />)
    const writeButton = getByRole('button', { name: 'Write' })

    fireEvent.click(writeButton)

    const textbox = queryByTestId('textbox')
    const markdown = queryByTestId('markdown')
    expect(markdown).toBeNull()
    expect(textbox).toBeTruthy()
    expect(container).toMatchSnapshot()
  })

  test('Should switch to Preview mode when user clicks Preview button', () => {
    const { container, getByRole, queryByTestId } = render(<MdInput />)
    const previewButton = getByRole('button', { name: 'Preview' })

    fireEvent.click(previewButton)

    const textbox = queryByTestId('textbox')
    const markdown = queryByTestId('markdown')
    expect(textbox.classList.contains('d-none')).toBe(true)
    expect(markdown).toBeTruthy()
    expect(container).toMatchSnapshot()
  })

  test('Should save cursor state going to and back from preview', () => {
    render(<TestComponent />)
    const previewButton = screen.getByRole('button', { name: 'Preview' })
    const writeButton = screen.getByRole('button', { name: 'Write' })
    const textbox = screen.getByRole('textbox')

    userEvent.click(textbox)
    userEvent.type(textbox, 'Hello,{enter}World!')
    expect(textbox.value).toBe('Hello,\nWorld!')

    textbox.setSelectionRange(7, 13)
    userEvent.click(previewButton)
    expect(textbox.classList.contains('d-none')).toBe(true)

    userEvent.click(writeButton)
    userEvent.type(textbox, '{backspace}Tom!')
    expect(textbox.value).toBe('Hello,\nTom!')
  })
  test('Should automatically resize to fit content', () => {
    render(<TestComponent />)
    const textbox = screen.getByRole('textbox')
    expect(textbox.style.height).toBe('')
    userEvent.type(
      textbox,
      'Hello,{enter}{enter}{enter}{enter}{enter}{enter}Tom!'
    )

    // Only picks up '+2px' additional padding as JSDOM doesnt properly mock styles
    // but this proves the code was exercised and is good enough for unit test
    expect(textbox.style.height).not.toBe('')
  })
  test('Show not auto resize after user sets height', () => {
    render(<TestComponent />)
    const textbox = screen.getByRole('textbox')

    const spy = jest
      .spyOn(textbox, 'clientHeight', 'get')
      .mockImplementation(() => 200)
      .mockImplementationOnce(() => 100)

    expect(textbox.style.height).toBe('')
    userEvent.click(textbox)
    expect(textbox.style.height).toBe('200px')

    userEvent.click(textbox)
    expect(textbox.style.height).toBe('200px')
    userEvent.type(
      textbox,
      'Lots{enter}{enter}{enter}{enter}o{enter}{enter}{enter}{enter}{enter}lines'
    )
    expect(textbox.style.height).toBe('200px')
  })
})
