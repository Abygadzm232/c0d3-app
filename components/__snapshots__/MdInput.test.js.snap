// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MdInput Component Should display markdown preview text when switch to Preview mode with input 1`] = `
<div>
  <div>
    <div
      class="nav nav-tabs"
    >
      <div
        class="nav-item"
      >
        <a
          class="nav-link"
          data-rb-event-key="Write"
          href="#"
          role="button"
          style="color: rgb(142, 142, 142);"
        >
          Write
        </a>
      </div>
      <div
        class="nav-item"
      >
        <a
          class="nav-link active"
          data-rb-event-key="Preview"
          href="#"
          role="button"
          style="color: rgb(41, 41, 41);"
        >
          Preview
        </a>
      </div>
    </div>
    <span
      class="preview"
      data-testid="markdown"
      style="min-height: 500px;"
    >
      Some 
      <strong>
        Text
      </strong>
    </span>
    <hr />
    <textarea
      class="textarea d-none"
      data-testid="textbox"
      placeholder="Type something..."
      style="height: 2px;"
    >
      Some **Text**
    </textarea>
  </div>
</div>
`;

exports[`MdInput Component Should render text onto textbox when user types 1`] = `
<div>
  <div>
    <div
      class="nav nav-tabs"
    >
      <div
        class="nav-item"
      >
        <a
          class="nav-link active"
          data-rb-event-key="Write"
          href="#"
          role="button"
          style="color: rgb(41, 41, 41);"
        >
          Write
        </a>
      </div>
      <div
        class="nav-item"
      >
        <a
          class="nav-link"
          data-rb-event-key="Preview"
          href="#"
          role="button"
          style="color: rgb(142, 142, 142);"
        >
          Preview
        </a>
      </div>
    </div>
    <textarea
      class="textarea"
      data-testid="textbox"
      placeholder="Type something..."
      style="height: 2px;"
    >
      Javascript
    </textarea>
  </div>
</div>
`;

exports[`MdInput Component Should switch to Write mode when user clicks Write button 1`] = `
<div>
  <div>
    <div
      class="nav nav-tabs"
    >
      <div
        class="nav-item"
      >
        <a
          class="nav-link active"
          data-rb-event-key="Write"
          href="#"
          role="button"
          style="color: rgb(41, 41, 41);"
        >
          Write
        </a>
      </div>
      <div
        class="nav-item"
      >
        <a
          class="nav-link"
          data-rb-event-key="Preview"
          href="#"
          role="button"
          style="color: rgb(142, 142, 142);"
        >
          Preview
        </a>
      </div>
    </div>
    <textarea
      class="textarea"
      data-testid="textbox"
      placeholder="Type something..."
    />
  </div>
</div>
`;
